Adding workflow execTimeout and keepActive properties

tsurdilo opened the pull request:
Signed-off-by: Tihomir Surdilovic <tsurdilo@redhat.com>

**Many thanks for submitting your Pull Request :heart:!**

**Please specify parts this PR updates:**

- [x ] Specification
- [x ] Schema
- [ ] Examples
- [ ] Extensions
- [ ] Roadmap
- [ ] Use Cases
- [ ] Community
- [ ] TCK
- [ ] Other

**What this PR does / why we need it**:
Adds the workflow "execWindow" . This property defines execution limitations of a single workflow instance.
Adding "adhoc" prop for start definitions
Adding "core concepts" sections to specification document

**Special notes for reviewers**:

**Additional information (if needed):**

ricardozanini commented on specification.md:
````diff
@@ -382,6 +397,112 @@ a resource containing an array of [retry](#Retry-Definition) definition.
+```
+
+Without `execWindow` it would be pretty hard to solve this business problem with control-flow logic alone.
+With it however it becomes pretty easy. Let's look at our sample workflow definition:
````
```suggestion
With it, however, it becomes pretty easy. Let's look at our sample workflow definition:
```

ricardozanini commented on specification.md:
````diff
@@ -382,6 +397,112 @@ a resource containing an array of [retry](#Retry-Definition) definition.
+}
+```
+
+Here we define the `execWindow` `time` property to "PT1H" meaning that after 1 hour of our workflow instance 
+collecting living room temperature data, we want to stop collecting data, and build our report by invoking the 
+"sendlivingroomtempreport" sub-workflow, before terminating the workflow instance. After the defined 
+amount of time, the very next living room temperature event will create a new instance of our workflow
+which will then again start collecting these events for 1 hour, build the report and terminate its execution, and so on.
````
```suggestion
Here we define the `execWindow` `time` property to "PT1H". This means that after 1 hour of the workflow instance collecting living room temperature data, we want to stop collecting it. Then we build the report by invoking the 
"sendlivingroomtempreport" sub-workflow before terminating the workflow instance. 

After the defined amount of time, the very next living room temperature event will create a new instance of the workflow
which will then again start collecting these events for 1 hour, build the report and terminate its execution, and so on.
```
I've found this paragraph too long and hard to understand the first time I read it. Not sure if I got it right, but I tried to keep the sentences smaller. wdyt?

manuelstein commented:
This flow control is pretty neat, but I think it could use a bit more text around the example.

How are temperature data collected? Is the instance looping an event state to consume events or does each temperature reading invoke the workflow? (see uniqueness of a workflow #146 ) in terms of the example: what if there's not just one living room temperature, but other rooms' temperature sensors that should use the same workflow, but concurrent instances for each room?

Also, after the time interval, a report should be created and sent by another workflow. The property to invoke this is a workflowId, but it's unclear if the current state is carried over to invoke it. Also, it says `Id of a workflow to be executed before workflow instance is terminated`, so if the window is not interrupting, then it sounds as if the next workflow would be triggered _while_ the instance is still executing, i.e. `before workflow instance is terminated`.

What starts the workflow?

What happens during the time the report is being generated? The old instance still exists (hasn't been terminated), but is no longer functional.

tsurdilo commented:
@manuelstein yeah i will add a proper full example in our examples markdown very soon, then things will be much more clear.

To answer some of your questions inline:
**How are temperature data collected?**: most likely case is that sensor data is in form of CE when it reaches the workflow end. Note that there has to be event correlation, so each event for a specific room is correlated to the same workflow instance.

**What if there's not just one living room temperature, but other rooms' temperature sensors that should use the same workflow** - this is just a small / simple example, yes you can correlate events from multiple sensors / sources with correlation (on the room id for example)

**What starts the workflow?** events

**What happens during the time the report is being generated? The old instance still exists (hasn't been terminated), but is no longer functional.** - in 100% of cases execution of a workflow instance is not async! So yes that workflow instance will remain active until its finished (report done) note that depending on the interrupting parameter this determines if during that time new events may still be processed or not..this depends on the use case, in this uses no new events should be processed.

**then it sounds as if the next workflow would be triggered while the instance is still executing**: I don't fully understand that paragraph but for this part of it alone: yes, there can be multiple instances of a workflow :)

jorgenj commented:
I see the usefulness here, but I think the context provided is focusing on events state use-cases, where-as I think we should think broadly about this as a 'global timeout for a workflow instance' feature.  

Imho, max-runtime or global-timeout, or execution-timeout seem to make more sense as a name for this attribute that 'execution window'.  ExecWindow to me implies that this would limit when a workflow is allowed to be invoked, rather than for how long it is expected to be running for.

In this PR, it looks like the intent is for a secondary workflow to be invoked when the execWindow has elapsed.  Some things that aren't clear to me:
1) What happens to the main workflow?  Does it just end at this point, or does it also transition at the point where the window has elapsed?  
2) Passing data; it's not clear to me, what data will be passed to that sub-workflow (sendlivingroomtempreport, in the example).  How will the user transform/filter what data should be passed to it?

tsurdilo commented:
@jorgenj 

1) after the subworkflow referenced by workflowId completes, the workflow instance is terminated
2) the current workflow data. it can be filtered by states of the subworkflow thats references

this change deals with a single workflow instance, so "global" maybe confusing. "execution" to me at least makes more sense as workflow execution happens inside a workflow instance. I do like "executionTimeout" what you suggested. Will update

tsurdilo commented:
rebased

jorgenj commented on schema/workflow.json:
```diff
@@ -199,12 +205,31 @@
+    "execwindow": {
+      "type": "object",
+      "properties": {
+        "time": {
```
Nit; maybe call this 'duration' instead?

jorgenj commented on schema/workflow.json:
```diff
@@ -199,12 +205,31 @@
+        },
+        "interrupt": {
+          "type": "boolean",
+          "description": "If `false`, workflow instance is allowed to finish current execution. If `true`, current workflow execution is abrupted.",
```
s/abrupted/stopped immediately/?

jorgenj reviewed (commented):
> @jorgenj
> 
> 1. after the subworkflow referenced by workflowId completes, the workflow instance is terminated
> 2. the current workflow data. it can be filtered by states of the subworkflow thats references

This seems reasonable to start with maybe, but I worry that this puts the onus on the subWorkflow to handle all possible inputs from all workflows that might use it, consider the following situation.
1) Customer has workflows p1, p2 & s1
2) p1 & p2 are top-level workflows which have an executionTimeout set
3) Both p1 & p2 re-use s1 when the executionTimeout is breached

This would mean that s1 must put in the effort of handling all possible inputs from both p1 & p2 which could be very difficult to maintain as the number of workflows grows.   Every time a new workflow is added that re-uses s1, the s1 implementation would have to be updated to deal with any differences in the state data of that new parent workflow.

Imho, it'd be better if we provided a way for the parent workflows to filter the data that they send to s1 so that users could try to make all of their parent workflows send the same structure, and s1 can focus on doing it's business logic rather than how to deal with the different types of state data it can receive from all the different workflows that use it.  @tsurdilo wdyt?

> this change deals with a single workflow instance, so "global" maybe confusing. "execution" to me at least makes more sense as workflow execution happens inside a workflow instance. I do like "executionTimeout" what you suggested. Will update

👍

tsurdilo commented on specification.md:
```diff
@@ -381,6 +396,112 @@ a resource containing an array of [retry](#Retry-Definition) definition.
+The `time` property defines the total amount of time in the execution window. Once a workflow instance is created,
+and the amount of the defined time is reached, the workflow instance needs to be terminated.
+
+The `interrupt` property defines if the currently running instance should be allowed to finish its current 
```
for error handling/compensation: yes, that is considered also workflow logic imo so should be interrupted if true
for interrupt being false the control flow logic should be allowed to be completed fully

tsurdilo commented on specification.md:
```diff
@@ -381,6 +396,112 @@ a resource containing an array of [retry](#Retry-Definition) definition.
+The `interrupt` property defines if the currently running instance should be allowed to finish its current 
+execution flow before it needs to be terminated. If set to `true`, the current instance execution must stop immediately.
+
+The `workflowId` property defines an unique id of a sub-workflow which should be executed before the workflow instance
```
yes, the workflow logic defined in subworkflow defines what should be done after exectimeout is reached

tsurdilo commented:
updated per recommendations
@jorgenj will work on an additional example per your suggestion and update pr so we can take a look at it next meet

tsurdilo commented on specification.md:
```diff
@@ -381,6 +396,112 @@ a resource containing an array of [retry](#Retry-Definition) definition.
+The `workflowId` property defines an unique id of a sub-workflow which should be executed before the workflow instance
+is terminated.
+
+`ExecWindow` allows us to solve interesting use cases, especially within the IoT (Internet of Things) domain.
```
@jorgenj i am working on additional use case involving batch processing where this can be used to cancel a "job" after a certain amount of time. hope this will cover the concern as well

tsurdilo replied:
added two new examples, added a very similar one as you requested but instead of dealing with patient registration multi-step process its an order confirmation multi-step process. hope this is ok

tsurdilo commented:
@ricardozanini @jorgenj @manuelstein: added a lot of explanations around what is being added with this pr in the spec doc. Also added two new examples that showcase the added features. 
You can see the added examples parsed here: https://github.com/tsurdilo/specification/blob/workflowtimeout/examples/examples.md
(look at the last two)

Please take a look again and let me know if there are still any concerns.

alankm commented:
Wouldn't it be a better design for `ExecTimeout` to let the user define an error to throw, rather than this `interrupt` and `runBefore` stuff? Then the implementation can just throw this error when the timeout is reached and it's up to the workflow to decide whether or not to terminate or "run before" using `onErrors`. Just like optionally defining a SIGINT handler in traditional programs.

Also, why is any of this `keepActive` "adhoc" stuff necessary? Loops can already be created trivially using transitions that point back to states that have appeared earlier in the workflow. As an example, I tested Tic-Tac-Toe implemented in a workflow just to see if I could:

```yaml
id: tictactoeGame
name: "Tic-Tac-Toe Game"
description: "Run a game of tic-tac-toe."
functions:
- name: PlayGameTurn
  metadata:
    image: vorteil/tictactoe
events:
- name: TurnPlayed
  type: tictactoeTurnPlayed
  source: tictactoeGame 
states:
- name: StartGame
  type: inject 
  data:
    game: ""
    player: 1
    board: [0, 0, 0, 0, 0, 0, 0, 0, 0]
    winner: ""
  start:
    kind: default 
  transition:
    nextState: RequestPlayerTurn
- name: RequestPlayerTurn 
  type: subflow 
  waitForCompletion: false 
  workflowId: tictactoePublish 
  transition: 
    nextState: ReceivePlayerTurn
- name: ReceivePlayerTurn 
  type: switch 
  eventTimeout: 1H
  eventConditions:
  - eventRef: TurnPlayed
    transition:
      nextState: ComputerTurn
  default:
    transition:
      nextState: AbortGame
- name: ComputerTurn 
  type: operation 
  actions:
  - functionRef:
      refName: PlayGameTurn
      parameters:
        game: "{{ $.game }}" 
        winner: "{{ $.winner }}" 
        player: "{{ $.player }}" 
        board: "{{ $.board }}" 
  transition:
    nextState: CheckForGameEnd
- name: CheckForGameEnd 
  type: switch 
  dataConditions:
  - condition: "$[?(@.winner=='player')]"
    transition:
      nextState: GameEnd
  - condition: "$[?(@.winner=='draw')]"
    transition:
      nextState: GameEnd
  - condition: "$[?(@.winner=='flux')]"
    transition:
      nextState: GameEnd
  default:
    transition:
      nextState: RequestPlayerTurn
  stateDataFilter:
    dataOutputPath: "$.after"
- name: AbortGame 
  type: inject 
  data:
    winner: "aborted"
  transition:
    nextState: GameEnd
- name: GameEnd
  type: subflow 
  waitForCompletion: false 
  workflowId: tictactoePublish 
  end:
    kind: default
```

The only thing here I can understand adding to the spec is a global timeout.

tsurdilo commented:
@alankm 
1) Regarding expressing with error handling: - no, as you do not know which part of your workflow will be active when execTimeout is triggered. So you rather in a large workflow explicitly define this in every single workflow state? 
Also this has nothing to do w/ errors. Errors represent actual business issues which are in SW declarative (explicitly defined in the workflow definition).
This is more of an instance-level timeout, nothing to do w/ errors, but its still explicitly defined. Runtimes impl can vary of course as how you want to define this internally.

2) Regarding loops - this has nothing to do with loops really. Yes there could be multiple ways to model some business requirements. This has to do with workflow instances. You might want in some cases to have workflow instances which just hold data. You cannot do that with loops. This has many use cases that go beyond looping.

alankm commented:
@tsurdilo "So you rather in a large workflow explicitly define this in every single workflow state?"
Definitely. Different behaviour is likely to be necessary based on what state the workflow is currently in. 

"This has many use cases that go beyond looping."
Can you provide a few different examples? The two you've added are trivial to implement with the global timeout being the only necessary addition. Even @jorgenj's example suggestion relies only on this global timeout. 

We shouldn't add more and more ways to model the same business requirements just because we can. 

Even collecting/holding data can be handled easily using only what's already in master. But I'm ignoring that because the topic of data collection seems like an anti-pattern to me -- something beyond the scope of serverlessworkflows now and forever. I certainly hope so, anyway.

tsurdilo commented:
@alankm sorry this is not how you express requirements to a workflow language in a domain specific way. The requirement "after 30 minutes of execution collecting data and create a report" is **much** better expressed this way than putting an error handler on 100 workflow states (and even subflows which are reusable and thus using errors can make them specific to this the parent instance if used as you mentioned).

Regarding "We shouldn't add more and more ways to model the same business requirements just because we can", i agree with this general statement, but at the same time allowing users to model business requirements in easier way is not a bad thing. I understand his might not present a valid use cases for yourself, but does not mean it will not be useful by others.

Regarding  "something beyond the scope of serverlessworkflows now and forever" - I do not understand this. SW states that its use are both stateless as well as stateful workflows. This may not have use for stateless workflows/orchestration but has very important use for stateful orchestration. If you are worried about memory or something, not sure this is an issue as for stateful workflow the runtime in most cases would not hold instances in memory, but persist them anyways and resume from file/disc/db when they need to be resumed.

tsurdilo commented:
@alankm and off topic -  do you have this tic tac toe workflow running? Would make a really cool demo for community!

tsurdilo commented:
@alankm here is the image of your presented workflow generated in https://serverlessworkflow.io/editor.html (or you can use the vscode extension to see the same) - very nice!

<img width="948" alt="Screen Shot 2021-01-17 at 6 10 10 PM" src="https://user-images.githubusercontent.com/119422/104859277-199d4800-58f2-11eb-8a2a-c06dff0c3c85.png">


I do see your point better now i think, for your case you can keep the state of one single game while that game is active. When the game ends you will loose the context of that game, and start a new game with a new workflow instance.

KeepAlive deals with a little bit of a different business problem: let's say you have your game workflow deployed 
in 3 different casinos. Your business problem is that you want to collect game data from each game in each casino and produce daily reports to your manager. 
You can create a workflow which has keepAlive set to true, which would receive events containg single game data and casino location for each game played. This would allow you to use this alongside execTimeout to create your daily/monthly/whatever reports. 
This present a pure workflow solution to the problem. Yes in your case you can say no thanks i can just persist this data and do the reports that way, but that's not the point of what we are doing. We are writing a workflow language and as such our goal is to provide solutions for problems expressed as workflows where there is a possibility. 

You can still persist your data if you want as part of the workflow, but the actual representation of your business requirements is contained in the workflow language and not 2,3, 100 different places. Hope this helps.

manuelstein commented:
The motivational use case has 2 workflows - one is the event collection, which starts and terminates for each event and the other one is a cron-triggered report generation. Putting both workflows in a single workflow definition is only necessary, because the report generation wants to access workflow data of the event collection workflow. This no longer describes a workflow, but an application. The solution to shift control between the two workflows assumes a computation model of a process with shared data (workflow data) among all individual event collections and the report generation (after which it forcefully terminates the instance, btw). But when I get 10k events/sec, I want to be able to scale this to parallel instances of the workflow. This is no longer possible when the workflow language prescribes/controls the termination of the workflow instance. In fact, it shouldn't contain any keep-alive semantic beyond having reached an end state. Instead, it should be possible to re-enter a previous application context.

IMHO, it's okay to want to share application data for different workflows. It may even be possible to allow multiple flows in a single state chart by collating the states into a single workflow definition. But there should be no assumption built-in on the process model.
We first have to clarify the Uniqueness constraint for workflows #146.

tsurdilo commented:
@manuelstein:
**This no longer describes a workflow, but an application.** - what does this mean? A workflow is a representation for a solution of your business requirements. Out of workflows you can build services...yes.

**This is no longer possible when the workflow language prescribes/controls the termination of the workflow instance. ** - I would argue that it **has to be**, how would you otherwise have portability? The runtimes can define custom exec but the outcome should be same as expressed in the workflow language. 

**it shouldn't contain any keep-alive semantic beyond having reached an end state. Instead, it should be possible to re-enter a previous application context.**- i dont understand what this means. each workflow instance contains its own context

 **We first have to clarify the Uniqueness constraint for workflows #146.**: this pr introduced workflow definition, workflow instance and clearly describes the lifetime of the instance. it is the starting point of that conversation and not the other way around

manuelstein commented:
By application, I mean it defines more than one workflow in a single document. One is triggered by an Event and the other one is triggered by a timeout. When no event is being received, the example assumes an ongoing execution, hence the "execTimeout". But I would want my workflows to persist the workflow data and go full serverless (i.e. remove all my running processes, no resource consumption) until I receive the next event. So, in my Serverless realization, there's no application process running that could perceive an execution timeout.

The workflow specification should make no assumptions about the engine implementation. I find it too limiting to specify that there should be a continuing execution while the workflow definition is not in any of its designed states.

A lot is currently hinting that a workflow instance is a running process. But if two requests/events correlate to the same instance, then it should be possible for the platform to persist the workflow context between the two event processings.

Yes, this is the starting point of the workflow instance lifecycle, but I think we should clarify first what triggers a new instance vs. what reuses an existing instance. E.g. when would a scheduled start state reuse an instance vs. create a new instance. How do I say that during one day, I want my hourly executions to reuse the old context and on the next day start fresh?
I think this would play well into this example with report generation. To me, reporting is a cron-like execution and it could be well coupled with an event collection by using two starting states in the same workflow document.

tsurdilo commented:
@manuelstein ok ..i no longer have the energy to exaplain this pr any more.... 
from my part LGTM ...and the is another LGTM from maintainer......

tsurdilo commented:
@manuelstein @ricardozanini so what we do with this pr? commit it / keep it around to discuss more in meetings / close it / wait for more discussions here / break it up into smaller prs / anything else ?
I think this pr brings a lot of improvements to our specification. If needed i can create more examples and demos around it so value is seen. Just would like to know what the end-result from our last meeting was on this. Please let me know

ricardozanini commented:
+1 to merge it, and I believe that we can always discuss this topic in our meetings to improve it. Others can also suggest modifications by sending new PRs around this topic.

tsurdilo commented:
@manuelstein ^ are you ok with this as well? I think we should move forward with this and start working on the expression language discussions in our meets.

tsurdilo commented:
@manuelstein any inputs? I'd like to move this forward please.

tsurdilo commented:
rebased

tsurdilo commented:
@manuelstein @ricardozanini @jorgenj 
Imma go ahead and more forwards with this. Understand that there might still be some open questions etc but there 
were no suggestions on how to change this stuff and we can always revise these sections before next release if needed.
So if you have any outstanding issues with this please open an issue and/or submit PRs against the spec/schema and we can get those in asap.
I just cant sit on this and have to constantly rebase these big prs its just taking way too much time.

manuelstein commented:
Workflow instance lifecycle will be the topic next Monday, Feb 1st, in our weekly call, @jorgenj @alankm @tsurdilo. I'll sum up what we discussed when we talked about this PR.

I'm also okay to have it merged now. Not that I'd agree with the functionality ;-) But merging it now will make it easier to work on. Otherwise we'd have to draft a change that has to consider a lot of different hanging PRs.

tsurdilo commented:
@manuelstein I agree to have discussions about this if you guys wish, but please make sure that discussion points are  related to some parts of this pr that need more explanation for example.
I think that if ppl want to discuss some generic topics around this it's much better to first do a PR about that and then discuss it in a meeting. I do not want to go off into some directions that are not related to a specific PR so everyone can relate.

manuelstein commented:
Got your point
